src/unit.cpp

```diff
@@ -3,96 +3,133 @@
 /// @author Simon Heybrock
 /// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
 /// National Laboratory, and European Spallation Source ERIC.
+#include "unit.h"
 #include <stdexcept>
 
-#include <boost/units/systems/si/amount.hpp>
-#include <boost/units/systems/si/dimensionless.hpp>
-#include <boost/units/systems/si/length.hpp>
-#include <boost/units/systems/si/area.hpp>
-
-#include "unit.h"
+// Helper type for the visitor id()
+template <class... Ts> struct overloaded : Ts... { using Ts::operator()...; };
+template <class... Ts> overloaded(Ts...)->overloaded<Ts...>;
+// Helper to check whether type is a member of a given std::variant
+template <typename T, typename VARIANT_T> struct isVariantMember;
+template <typename T, typename... ALL_T>
+struct isVariantMember<T, std::variant<ALL_T...>>
+    : public std::disjunction<std::is_same<T, ALL_T>...> {};
 
-Unit makeUnit(const boost::units::si::dimensionless &) {
-  return {Unit::Id::Dimensionless};
+/// Construct unit from a given Id
+Unit::Unit(const Unit::Id id) {
+  switch (id) {
+  case Unit::Id::Dimensionless:
+    m_unit = boost::units::si::dimensionless();
+    break;
+  case Unit::Id::Length:
+    m_unit = boost::units::si::length();
+    break;
+  case Unit::Id::Area:
+    m_unit = boost::units::si::area();
+    break;
+  case Unit::Id::AreaVariance:
+    m_unit = boost::units::si::area() * boost::units::si::area();
+    break;
+  case Unit::Id::Counts:
+    m_unit = neutron::tof::counts();
+    break;
+  case Unit::Id::CountsVariance:
+    m_unit = neutron::tof::counts() * neutron::tof::counts();
+    break;
+  case Unit::Id::InverseLength:
+    m_unit = boost::units::si::dimensionless() / boost::units::si::length();
+    break;
+  case Unit::Id::Energy:
+    m_unit = neutron::tof::energy();
+    break;
+  case Unit::Id::Wavelength:
+    m_unit = neutron::tof::wavelength();
+    break;
+  case Unit::Id::Time:
+    m_unit = boost::units::si::time();
+    break;
+  case Unit::Id::Tof:
+    m_unit = neutron::tof::tof();
+    break;
+  case Unit::Id::Mass:
+    m_unit = boost::units::si::mass();
+    break;
+  }
 }
-Unit makeUnit(const boost::units::si::length &) { return {Unit::Id::Length}; }
-Unit makeUnit(const boost::units::si::area &) { return {Unit::Id::Area}; }
-Unit makeUnit(const decltype(std::declval<boost::units::si::amount>() *
-                             std::declval<boost::units::si::amount>()) &) {
-  return {Unit::Id::CountsVariance};
+
+/// Get the Id corresponding to the underlying unit
+Unit::Id Unit::id() const {
+  // return Unit::Id::Dimensionless;
+  return std::visit(
+      overloaded{[](boost::units::si::dimensionless) {
+                   return Unit::Id::Dimensionless;
+                 },
+                 [](boost::units::si::length) { return Unit::Id::Length; },
+                 [](boost::units::si::area) { return Unit::Id::Area; },
+                 [](decltype(std::declval<boost::units::si::area>() *
```

Comment:
@SimonHeybrock do you think i should also use the short variable names here to have something like `decltype(m2 * m2)` ?

Reply:
Yes, that would probably make sense.